### PR TITLE
Implement remaining value sanitation algorithms

### PR DESCRIPTION
**Pull request by lanmonster**
left a couple FIXMEs. it looks like the `multiple` attribute is not implemented yet, so I did not handle that case for the email type, and also sometimes there are 53 weeks in a year.

### CODE REVIEW

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 1]
```diff
@@ -497,6 +497,18 @@ String HTMLInputElement::value_sanitization_algorithm(String value) const
             return builder.to_string();
         }
     } else if (type_state() == HTMLInputElement::TypeAttributeState::URL) {
+        // Strip newlines from the value, then strip leading and trailing ASCII whitespace from the value.
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:698]: `// https://html.spec.whatwg.org/multipage/input.html#url-state-(type=url):value-sanitization-algorithm`

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 2]
```diff
@@ -518,7 +530,7 @@ String HTMLInputElement::value_sanitization_algorithm(String value) const
         // If the value of the element is a valid simple color, then set it to the value of the element converted to ASCII lowercase;
         if (is_valid_simple_color(value))
             return value.to_lowercase();
-        
+
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:757]: Don't fix whitespace across commits

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 3]
```diff
@@ -483,6 +483,90 @@ static bool is_valid_simple_color(String const& value)
...
+        return false;
+    if (!is_ascii_digit(parts[1][1]))
+        return false;
+
+    auto month = (parse_ascii_digit(parts[1][0]) * 10) + parse_ascii_digit(parts[1][1]);
+    return month >= 1 && month <= 12;
+}
+
+// https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#number-of-days-in-month-month-of-year-year
+static u32 number_of_days_in_month_year(u32 month, u64 year)
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:612]: Let's use the existing `AK::days_in_month()` for this

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 4]
```diff
@@ -508,7 +592,7 @@ String HTMLInputElement::value_sanitization_algorithm(String value) const
         }
     } else if (type_state() == HTMLInputElement::TypeAttributeState::Email) {
         // FIXME: handle the `multiple` attribute
-        
+
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:709]: Don't fix whitespace across commits

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 5]
```diff
@@ -525,6 +609,10 @@ String HTMLInputElement::value_sanitization_algorithm(String value) const
         auto maybe_double = value.to_double(TrimWhitespace::Yes);
         if (!maybe_double.has_value() || !isfinite(maybe_double.value()))
             return "";
+    } else if (type_state() == HTMLInputElement::TypeAttributeState::Date) {
+        // https://html.spec.whatwg.org/multipage/input.html#date-state-(type=date):value-sanitation-algorithm
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:727]:
```suggestion
        // https://html.spec.whatwg.org/multipage/input.html#date-state-(type=date):value-sanitization-algorithm
```

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 6]
```diff
@@ -613,6 +613,10 @@ String HTMLInputElement::value_sanitization_algorithm(String value) const
         // https://html.spec.whatwg.org/multipage/input.html#date-state-(type=date):value-sanitation-algorithm
         if (!is_valid_date_string(value))
             return "";
+    } else if (type_state() == HTMLInputElement::TypeAttributeState::Month) {
+        // https://html.spec.whatwg.org/multipage/input.html#month-state-(type=month):value-sanitation-algorithm
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:731]:
```suggestion
        // https://html.spec.whatwg.org/multipage/input.html#month-state-(type=month):value-sanitization-algorithm
```

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 7]
```diff
@@ -617,6 +662,10 @@ String HTMLInputElement::value_sanitization_algorithm(String value) const
         // https://html.spec.whatwg.org/multipage/input.html#month-state-(type=month):value-sanitation-algorithm
         if (!is_valid_month_string(value))
             return "";
+    } else if (type_state() == HTMLInputElement::TypeAttributeState::Week) {
+        // https://html.spec.whatwg.org/multipage/input.html#week-state-(type=week)
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:735]:
```suggestion
        // https://html.spec.whatwg.org/multipage/input.html#week-state-(type=week):value-sanitization-algorithm
```

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 8]
```diff
@@ -666,6 +715,10 @@ String HTMLInputElement::value_sanitization_algorithm(String value) const
         // https://html.spec.whatwg.org/multipage/input.html#week-state-(type=week)
         if (!is_valid_week_string(value))
             return "";
+    } else if (type_state() == HTMLInputElement::TypeAttributeState::Time) {
+        // https://html.spec.whatwg.org/multipage/input.html#time-state-(type=time)
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:739]:
```suggestion
        // https://html.spec.whatwg.org/multipage/input.html#time-state-(type=time):value-sanitization-algorithm
```

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 9]
```diff
@@ -719,6 +739,11 @@ String HTMLInputElement::value_sanitization_algorithm(String value) const
         // https://html.spec.whatwg.org/multipage/input.html#time-state-(type=time)
         if (!is_valid_time_string(value))
             return "";
+    } else if (type_state() == HTMLInputElement::TypeAttributeState::LocalDateAndTime) {
+        // https://html.spec.whatwg.org/multipage/input.html#local-date-and-time-state-(type=datetime-local)
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:743]:
```suggestion
        // https://html.spec.whatwg.org/multipage/input.html#local-date-and-time-state-(type=datetime-local):value-sanitization-algorithm
```

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 10]
```diff
@@ -744,6 +744,11 @@ String HTMLInputElement::value_sanitization_algorithm(String value) const
         if (is_valid_local_date_and_time_string(value))
             return normalize_local_date_and_time_string(value);
         return "";
+    } else if (type_state() == HTMLInputElement::TypeAttributeState::Range) {
+        // https://html.spec.whatwg.org/multipage/input.html#range-state-(type=range)
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:748]:
```suggestion
        // https://html.spec.whatwg.org/multipage/input.html#range-state-(type=range):value-sanitization-algorithm
```

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 11]
```diff
@@ -489,16 +704,58 @@ String HTMLInputElement::value_sanitization_algorithm(String value) const
             }
             return builder.string_view().trim(Infra::ASCII_WHITESPACE);
         }
+    } else if (type_state() == HTMLInputElement::TypeAttributeState::Email) {
+        // FIXME: handle the `multiple` attribute
+        // Strip newlines from the value, then strip leading and trailing ASCII whitespace from the value.
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:709]:
```suggestion
        // https://html.spec.whatwg.org/multipage/input.html#email-state-(type=email):value-sanitization-algorithm
        // Strip newlines from the value, then strip leading and trailing ASCII whitespace from the value.
```

**Review by linusg — APPROVED** [review]
One missing link, rest looks great :)

**Review by linusg — APPROVED** [review]

**Comment by AtkinsSJ** [comment]
CI is failing to build.

**Pull request merged**